Avoid relying on package name to parse CraftBukkit version.

portlek opened the pull request:
Fix #294 

the pr might be a bit confusing. here are the things done:
- revision names from wrapper class names removed.
- searches the version wrapper starting from the server versions' patch version to 0.
- move the special version classes to their own modules.
- implemented a simple version parser using Bukkit#getVersion().

tested on 1.8.6, 1.8.8, 1.10.2, 1.17.1.

portlek commented:
> What's the point of still having the R-Version in the module name while removing them in the class names? This is inconsistent.

Although I agree with what you are saying, but I think it should stay the same because if we change the module names, every single artifactId in the maven repository will change and that might not be what @WesJD wants. It's up to him tho.

> Also, do we really need all these new classes in the API module? Afaik you copied them partially from the PaperLib, why can't they be shaded in? It's not ideal to have unattributed code from others in the repo. As we discussed in #294, we don't even need a lib for this. Getting the minor, major and patch version is really just parsing the version string.

The reason why I didn't shade the paperlib in is because it's containing too many unnecessary features in it. I just copied the necessary part which parses the Minecraft version using Bukkit#getVersion() method, which is basically what you are saying here:

> Getting the minor, major and patch version is really just parsing the version string.

Thanks for the feedback on this.

mastercake10 commented:
> every single artifactId in the maven repository will change

Why is this a problem? The modules get shaded in the final jar anyway, I don't think people use the Wrapper classes explicitly in their plugins. Even then they can still rename the module.

I think if we're going to ditch the R-Version, we should do it right, which means renaming the packages too.

> The reason why I didn't shade the paperlib in is because it's containing too many unnecessary features in it. I just copied the necessary part which parses the Minecraft version using Bukkit#getVersion() method, which is basically what you are saying here:

Do you think we can use their code without attribution? I am not an expert when it comes to licenses, but it doesn't feel right. I think this can be done without using PaperLib at all. We don't even use the methods `Environment#isPaper` etc.

portlek commented:
remove almost every class and end up with a single version parser which makes things a lot simpler.

portlek commented on api/src/main/java/net/wesjd/anvilgui/version/VersionProvider.java:
```diff
@@ -0,0 +1,59 @@
+        int major = 1;
+        int minor = 0;
+        int patch = 0;
+        if (matcher.find()) {
```
Maybe I should put a warning or throw an exception to notify the user that using the library in completely different environment.

mastercake10 replied:
I don't know, are there any other "Environments" that use Bukkit and don't follow the `VERSION_PATTERN`?

portlek replied:
that's the thing I am not aware of.

WesJD replied:
Yeah, I think an exception should be thrown if it can't match. We need to give the devs/end users as much info as possible to debug their issue.

mastercake10 commented:
Looking good.

> It's an MIT license, which basically means that you do whatever you want with it. And plus, I've put the license file from the paperlib repository as well.

Yeah, I know. But it still needs attribution, which is basically what I was saying. The [LICENSE](https://github.com/forkstuffs/AnvilGUI/blob/versions/api/src/main/java/net/wesjd/anvilgui/version/LICENSE) file is empty btw. Mabye put it in the header of the file(s) so others know which parts are exactly taken from paper / licensed under MIT? I was thinking we could just avoid using PaperLib since its really not that hard to parse a simple string.

portlek commented:
> Yeah, I know. But it still needs attribution, which is basically what I was saying. The [LICENSE](https://github.com/forkstuffs/AnvilGUI/blob/versions/api/src/main/java/net/wesjd/anvilgui/version/LICENSE) file is empty btw. Mabye put it in the header of the file(s) so others know which parts are exactly taken from paper / licensed under MIT? I was thinking we could just avoid using PaperLib since its really not that hard to parse a simple string.

didn't notice it was empty. i've removed the license and everything related to paperlib so we should be good now.

mastercake10 commented:
> didn't notice it was empty. i've removed the license and everything related to paperlib so we should be good now.

Parts of the code are still from PaperLib, so it needs the attribution + license.

0dinD commented:
In regards to the renaming of the packages and Maven artifacts: I think it's fine, since most (if not all) users of AnvilGUI probably just depend on the main AnvilGUI artifact, not the individual adapter modules. So the rename will not affect them.

If we make the decision to keep the old "R-version" module/package names, I think we should also keep the old class names, for consistency. Then we would need to have a hybrid approach where we use the new naming scheme for 1.20.5+ and fall back to the old "R-version" naming scheme for older modules. But for what it's worth, if I had to choose, I would ditch the "R-versions" entirely and rename the packages and Maven artifacts.

---

In regards to the version parsing, I am quite convinced that we could just use `Bukkit.getServer().getBukkitVersion().split("-")[0]` instead of trying to parse `Bukkit.getVersion()`, as discussed in https://github.com/WesJD/AnvilGUI/issues/294#issuecomment-2087862734. The reason why PaperLib does that is because it gives a few more bits of information, such as information about whether the Minecraft version is a pre-release or release candidate. But I don't see how this information is important to us. And if we use `Bukkit.getServer().getBukkitVersion().split("-")[0]`, we can avoid copying from PaperLib at all and simplify the `VersionProvider` class.

0dinD commented on api/src/main/java/net/wesjd/anvilgui/version/VersionMatcher.java:
```diff
@@ -18,22 +16,40 @@ public class VersionMatcher {
+        // Start searching for suitable VersionWrapper
+        // starting from the current patch version going downward
+        VersionWrapper suitableWrapper = null;
+        for (int i = patch; i >= 0; --i) {
```
This would stop searching at the lowest patch version. Which works fine so far, because each new minor version thus far has required a new AnvilGUI module, due to Spigot's "R-version" packages.

But if Spigot stops using "R-version" packages in the future (which would not surprise me), we could end up in a situation where we have a Wrapper for, let's say, Minecraft 1.25, which also works for Minecraft 1.26. Then we would not be able to find the 1.25 Wrapper class, because the search stops at 1.26.

Honestly this is purely hypothetical at this point and doesn't really matter (we can fix it later if the problem arises), but I just wanted to highlight it for discussion.

The reason I bring this up is that I am wondering if it would be simpler (and also more future-proof) to just define an ordered list of Wrapper names (`1_8`, `1_8_3`, `1_8_8`, `...`, `1_20_5`) and then have the `VersionMatcher` try to load the highest supported version from the list. Just start from the back of the list and continue until a supported version is found (one that is less than or equal to the current version), then load that module. This is just an idea that came to my mind, let me know what you think.

portlek replied:
not quite sure what you are talking about. the lowest patch version would be zero, and if it's zero, I don't use the patch version, and instead I just use major_minor pattern. I really don't understand what can cause an issue here from what you are saying. because we have 1.10 and 1.11 versions as well. so what's the difference between 1.10–1.11 and 1.25–1.26? since we're going to separate 1.25 and 1.26 versions as WrapperV1_25 and WrapperV1_26 like how we did it in 1.10–1.11 versions.
the main reason why I don't like defining every single version in a map is because it'll require changing the map values on every single release. so I don't like putting new things that can go wrong while reviewing. making things as automatic as possible would be much more cleaner imo.

portlek replied:
> The problem with your current solution is that (unless I'm mistaken) it would start searching for `1_26_3`, then for `1_26_2`, `1_26_1` and finally for `1_26`. After that, it would give up the search and `throw new IllegalStateException("No compatible version wrapper found for server version v" + major + "." + minor + "." + patch);`. But that's not correct, because _there is_ a compatible version wrapper (`1_25_2`).

that's not a problem tho, we have to provide wrapper versions for each minecraft version. if there is no 1_26 wrapper which means we don't support 1.26 version yet. this is already how the library works right now. why we jump from 1_25 to 1_28? that doesn't make sense and never gonna happen either way.

0dinD replied:
> this is already how the library works right now. why we jump from 1_25 to 1_28? that doesn't make sense and never gonna happen either way.

That *can* happen in the future, if Spigot stops using the "R-versions" and just uses Mojang mappings (same as Paper), which wouldn't surprise me. When that happens, it is totally possible for a 1.25.2 wrapper to be compatible with Minecraft 1.26, 1.27 etc. This can already be done for Paper, as mentioned by @mastercake10 in #294:

> I see one huge advantage in method 2), which is that the new wrapper class can be re-used for all upcoming paper versions, since it doesn't rely on imports with cb version numbers in it. Also mojang mappings doesn't change over releases.

mastercake10 replied:
This is why we discussed just checking if the package name contains an R-Version, and if it does not, load the module through a lookup table in. We can't predict the future. As long as this project supports spigot, i don't really see the benefit of dropping the R-Version entirely. We haven't discussed it before, so I don't see why Paper's decision suddenly makes us doing this. We really just need to make it compatible with Paper again. This is also the consensus we reached in #294.

There is so much discussion going on here, we simply don't know what Spigot's next moves are, so why not just take the simple approach and see what the future brings?

portlek replied:
it doesn't make a difference if spigot is also remove revisions. we can always add support for new versions simply by adding more nms modules. there won't be any issue even if spigot decides to go down on the same road as paper.

portlek replied:
> That is not the same situation. There is a WrapperV1_10 and a WrapperV1_11, so your current solution will work for these versions because the search never has to continue down to a lower minor version. As illustrated above, the hypothetical case I described above is different in that there is a WrapperV1_25_2, which is also compatible with 1.26.x versions, but there are no WrapperV1_26_X version wrappers, so your current algorithm will not find a wrapper if the current version is 1.26.x even though there is a compatible one (WrapperV1_25_2).


if there is no 1_26_x version wrapper, then that means we don't support 1.26 yet, because we didn't create version wrapper for it?
i really don't understand what you guys are trying to say here. because if you think about it, if you add the version wrapper specifically for 1.26 then it will work, what's the problem here?

mastercake10 replied:
> it doesn't make a difference if spigot removes revisions. we can always add support for new versions simply by adding more nms modules. there won't be any issue even if spigot decides to go down on the same road as paper.

Right, that's why we don't even need to drop the R at this point.

> i really don't understand what you guys are trying to say here. because if you think about it, if you add the version wrapper specifically for 1.26 then it will work, what's the problem here?

It assumes that we have wrappers for mojang-mapped servers (which is the case for paper 1.20.5+). The new wrapper class for 1.26 _could_ exactly be the same file for 1.25, so it might be as well just be reused instead of adding the same file again. Of course with Spigot mappings, the method names etc. change every release, which is why we need to add a new wrapper for every MC version.

portlek replied:
so the problem is not the mappings or R or anything, just DIY?

portlek replied:
we can create a base class for it so that would be really easy solution.

portlek replied:
> Right, that's why we don't even need to drop the R at this point.

i think we can do that but this is not the PR for it. maybe after refactoring the build system to gradle.

mastercake10 replied:
> so the problem is not the mappings or R or anything, just DIY?

I don't think there is a problem for now, its just based on assumptions for the future. 

It comes down to personal preference, for me this whole process of finding the right Wrapper class for a given MC version is not really required, especially for < MC 1.20.5, where you can still look up the R-Version.

> i think we can do that but this is not the PR for it. maybe after refactoring the build system to gradle.

I didn't mean the module names, you already dropped the R in the class names. I have already a fork for gradle support as discussed [here](https://github.com/WesJD/AnvilGUI/issues/294#issuecomment-2082700651), but it doesn't drop the R (see the comment why).

portlek replied:
i think we should think about all of these whenever we need, let's just focus on merging this pr. sent request.

0dinD replied:
> There is so much discussion going on here, we simply don't know what Spigot's next moves are, so why not just take the simple approach and see what the future brings?

> i think we should think about all of these whenever we need,

Once again, I brought this stuff with `VersionMatcher` up only for discussion, I agree that this is not something we have to fix right now.

> This is why we discussed just checking if the package name contains an R-Version, and if it does not, load the module through a lookup table in.

> It comes down to personal preference, for me this whole process of finding the right Wrapper class for a given MC version is not really required, especially for < MC 1.20.5, where you can still look up the R-Version.

This is the second point that I was trying to make, which is that the current `VersionMatcher` implementation (apart from possibly breaking in the future if Spigot moves in the direction of Paper and uses Mojang mappings without "R-versions") seems unnecessarily complex when we just simply just have a lookup table or list.

---

Once again, I don't oppose the current `VersionMatcher` implementation, it will definitely work. Ultimately the decision is not up to me either way, I am not a maintainer here, just throwing my ideas around.

WesJD replied:
Good point @0dinD, but I think we should wait to fix this issue until the problem arises.

portlek commented:
> In regards to the version parsing, I am quite convinced that we could just use `Bukkit.getServer().getBukkitVersion().split("-")[0]` instead of trying to parse `Bukkit.getVersion()`, as discussed in [#294 (comment)](https://github.com/WesJD/AnvilGUI/issues/294#issuecomment-2087862734). The reason why PaperLib does that is because it gives a few more bits of information, such as information about whether the Minecraft version is a pre-release or release candidate. But I don't see how this information is important to us. And if we use `Bukkit.getServer().getBukkitVersion().split("-")[0]`, we can avoid copying from PaperLib at all and simplify the `VersionProvider` class.
> 
> But maybe there is some advantage to parsing `Bukkit.getVersion()` that I have missed?

Bukkit#getVersion() is much more simple to work with in terms of the knowledge about the way Bukkit works. The main reason why I think that is because Bukkit#getBukkitVersion() defined by Bukkit, but Bukkit#getVersion() method is using the thing that represents the Mojang's implementation. So if MinecraftServer#getServerVersion() implementation changes, which is highly unlikely, then we will know immediately and change the version pattern in the version provider. But let's say if some random fork decided to change CraftServer#getBukkitVersion(), we cannot catch that change easily since we cannot know every single spigot/paper fork on the planet.

0dinD commented:
> Bukkit#getVersion() is much more simple to work with in terms of the knowledge about the way Bukkit works. The main reason why I think that is because Bukkit#getBukkitVersion() defined by Bukkit, but Bukkit#getVersion() method is using the thing that represents the Mojang's implementation. So if MinecraftServer#getServerVersion() implementation changes, which is highly unlikely, then we will know that immediately and change the version pattern in the version provider. But let's say if some random fork decided to change CraftServer#getBukkitVersion(), we cannot catch that change easily since we cannot know every single spigot/paper fork on the planet.

Okay, that is a fair point. Although I honestly find it hard to believe that a fork would change the `getBukkitVersion` implementation that drastically. At that point, what's to say that they don't also change `getVersion` to return something in a different format as well? Like, neither of those methods actually make any guarantees about the return string format, as far as I know. But then again, I would agree that `getBukkitVersion` is more likely to change than `getVersion`, so honestly I'm fine with your solution.

0dinD commented:
Personally I would have liked to see either sticking entirely with the old "R-version" names for both the modules and the class names (which would lead to a very simple PR with minimal changes) or the complete opposite where we remove all traces of "R-versions", even from the module names. Currently, we have an inconsistent middle ground.

But for what it's worth, I still approve of the current solution, because it will work, and at the end of the day that's what really matters.

WesJD commented on api/src/main/java/net/wesjd/anvilgui/version/VersionProvider.java:
```diff
@@ -0,0 +1,85 @@
+            final MatchResult matchResult = matcher.toMatchResult();
+            try {
+                major = Integer.parseInt(matchResult.group(1), 10);
+            } catch (final Exception ignored) {
```
Why are these exceptions ignored? Isn't the version unable to be parsed correctly? Shouldn't a new error like `CannotParseMinecraftVersionException` or something be thrown?

WesJD reviewed (changes requested):
Overall, good work here. I really appreciate your help.

Since you're renaming the classes to remove the concept of an "R-version", then the module names should have the "R-version" removed from them too.

I think that this PR could have been much simpler (i.e. without touching all of the past versions), but it's not going to hurt anything to make this change. It just means more mental overhead to read the diff.

mastercake10 reviewed (approved):
Looking good now, but yeah dropping the R-Version on mc < 1.20.5 is not essentially needed, but shouldn't hurt either.

0dinD commented on 1_10_2/pom.xml:
The module name should be `1_10`, not `1_10_2`, to match the wrapper version

0dinD commented on 1_16_1/pom.xml:
The module name should be `1_16`, not `1_16_1`, to match the wrapper version

0dinD commented on api/src/main/java/net/wesjd/anvilgui/version/VersionProvider.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.regex.Pattern;
+import org.bukkit.Bukkit;
+
+/**
+ * Partially copied from <a href="https://github.com/PaperMC/PaperLib/blob/master/src/main/java/io/papermc/lib/environments/Environment.java#L56">here</a>.
+ */
```
Would be good to use a permalink to a specific commit instead of the master branch (which can change in the future and make the link dead). A permalink such as: https://github.com/PaperMC/PaperLib/blob/69e76fc9036e7b5f2c967bd21db31043724ef9e3/src/main/java/io/papermc/lib/environments/Environment.java#L56

0dinD commented on 1_10/pom.xml:
```diff
@@ -10,7 +10,7 @@
     </parent>
 
-    <artifactId>anvilgui-1_10_R1</artifactId>
+    <artifactId>anvilgui-1_10_2</artifactId>
```
The `artifactId` should be `1_10` instead of `1_10_2`, to match the module and Wrapper class name

0dinD commented on 1_16/pom.xml:
```diff
@@ -10,7 +10,7 @@
     </parent>
 
-    <artifactId>anvilgui-1_16_R1</artifactId>
+    <artifactId>anvilgui-1_16_1</artifactId>
```
The `artifactId` should be `1_16` instead of `1_16_1`, to match the module and Wrapper class name

0dinD commented on 1_14/pom.xml:
This dependency can be removed now, because the usage of `version.special.AnvilContainer1_14_4_R1` was removed: 

```xml
        <dependency>
            <groupId>net.wesjd</groupId>
            <artifactId>anvilgui-1_14_4_R1</artifactId>
            <version>${project.parent.version}</version>
            <scope>provided</scope>
        </dependency>
```

0dinD commented on 1_17/pom.xml:
This dependency can be removed now, because the usage of `version.special.AnvilContainer1_17_1_R1` was removed: 

```xml
        <dependency>
            <groupId>net.wesjd</groupId>
            <artifactId>anvilgui-1_17_1_R1</artifactId>
            <version>${project.parent.version}</version>
            <scope>provided</scope>
        </dependency>
```

0dinD commented on 1_19/pom.xml:
This dependency can be removed now, because the usage of `version.special.AnvilContainer1_19_1_R1` was removed: 

```xml
        <dependency>
            <groupId>net.wesjd</groupId>
            <artifactId>anvilgui-1_19_1_R1</artifactId>
            <version>${project.parent.version}</version>
            <scope>provided</scope>
        </dependency>
```

0dinD commented:
Looks like these dependencies have not been updated to use the new module names:

https://github.com/forkstuffs/AnvilGUI/blob/90fcd18557b98ac2a1a45efca509f2a4296a1bdd/api/pom.xml#L28-L195

0dinD reviewed (approved):
I have not had time to test this (run it on a server), but the code looks good to me now. Good work!

WesJD commented:
@portlek Can you confirm which versions you tested the latest commit on?

portlek commented:
> @portlek Can you confirm which versions you tested the latest commit on?

i didn't test the latest commits, whenever i get time. i will test on different versions.

WesJD commented:
@mastercake10 @0dinD @portlek could someone test the latest commit so we can get this out?

mastercake10 commented:
> @mastercake10 @0dinD @portlek could someone test the latest commit so we can get this out?

I think this should be fine,  I've tested it on a few MC versions. Shouldn't break anything.

But I have my [gradle fork](https://github.com/mastercake10/AnvilGUI/tree/no_relocation) that already has support for 1.20.5/6. If we decide to use Gradle instead of maven later on, all the changes of this PR would be useless since I haven't renamed the NMS modules.

The reason why I haven't created this PR is that I am not sure if the [PaperWrapper](https://github.com/mastercake10/AnvilGUI/blob/no_relocation/NMS/paper/src/main/java/net.wesjd.anvilgui.version/WrapperPaper.java) will stay the same for the next few minecraft versions, its just speculations at this point. 

This is why we proposed not to rename the modules in #294 and take a simple approach instead (something like just to check if the craft bukkit version contains an R, if not, get the minecraft version and find the appropriate "R-Module" through a look up table).

I don't want to void all the work @portlek put into this PR, I really appreciate it. Maybe we should find a middle ground and create a new PR without changing that much to make it work for 1.20.5/6? We can still merge this or my gradle fork when necessary later on.

WesJD commented:
I'm good for you to take the lead and make a decision there. If you think it's better to do a simpler PR, just whip one up. I just want to make sure we are on track to support the new versions, as I've been getting a lot of requests for it.

BySwiizen commented:
> Je suis bon pour que vous preniez les devants et que vous preniez une décision là-bas. Si vous pensez qu’il est préférable de faire un PR plus simple, il suffit d’en créer un. Je veux juste m’assurer que nous sommes sur la bonne voie pour prendre en charge les nouvelles versions, car j’ai reçu beaucoup de demandes pour cela.

I think it would be more favorable firstly to make a simple commit which adds the new versions and secondly, to create a branch with all the more technical modifications to be made.

BySwiizen commented:
> I'm good for you to take the lead and make a decision there. If you think it's better to do a simpler PR, just whip one up. I just want to make sure we are on track to support the new versions, as I've been getting a lot of requests for it.

I think it would be more favorable firstly to make a simple commit which adds the new versions and secondly, to create a branch with all the more technical modifications to be made.

WesJD commented:
I'm closing this since we just merged #326 . Let's make these technical changes in another PR if still desired.
